### PR TITLE
Implement vector finite elements

### PR DESCRIPTION
**Pull request by lindsayad**
Important changes:
  - Introduced a finite element variable base class called MooseVariableFE
    - Contains methods and data that don't deal with the interpolated solution (e.g.
      _u, _grad_u, _curl_u, etc.)
  - MooseVariableField (this name could be changed) is a class template that
    inherits from MooseVariableFE and implements the methods and data that deal
    with the interpolated solution data members.
    - I've typedef'd `MooseVariableField<Real>` to be the well-known MooseVariable
    - `MooseVariableField<VectorValue<Real>>` is typedef'd as MooseVariableVector
  - Some methods in Assembly have been templated so that they return the proper shape
    functions corresponding to scalar shape functions for MooseVariable and vector
    shape functions for MooseVariableVector
  - Refactored EigenKernel to inherit directly from Kernel instead of KernelBase
  - VectorKernel inherits from KernelBase
  - Created a new BoundaryCondition class called IntegratedBCBase
    - Traditional IntegratedBC is the scalar version that inherits from IntegratedBCBase
    - VectorIntegratedBC is the vector version that inherits from IntegratedBCBase
  - Added a test that mimics libmesh/examples/vector_fe/ex3
  - Implements `NEDELEC_ONE` but can very easily add `LAGRANGE_VEC` which at least @dschwen is interested in

There are undoubtedly some rough edges in here, and I should add a test for coupling a vector variable with a regular variable for example, but I figured I'd get this up here in case of discussion.

Closes (#10049)

### CODE REVIEW

**Comment by lindsayad** [comment]
The test mimics example 3 in libmesh's `examples/vector_fe` directory. Results:

### ux
ux

### uy
uy

tagging @cticenhour

**Comment by dschwen** [comment]
```
MooseVariable.h: No such file or directory
```

Breaks a ton of builds (Marmot for example)
Can you maybe add that file back in (even if it now just contains a single(?) include directive to include the new file)?

**Comment by lindsayad** [comment]
I like that fix. Will do

**Comment by gridley** [comment]
How far do you think this is from completion @lindsayad? I may go ahead and pull your branch to start P1 in Moltres. Also, is there a _div_u available? Just about to look at the code, but so many files appear to have been changed; it's hard to pinpoint where to look.

**Comment by andrsd** [comment]
> Also, is there a _div_u available?

I searched through the patch and found only `_curl_u`.

**Comment by andrsd** [comment]
I plan on looking at this PR after the holidays.

**Comment by lindsayad** [comment]
David's right; there's no div_u in this patch, but it would not be
difficult to add if it's desired.

**Comment by lindsayad** [comment]
inherit-graph

@friedmud Here's the inheritance graph for moose variables for this PR. I know the names aren't great, and I am open to suggestions for better. One of the reasons for having the `MooseVariableFE` class is so that I can have containers, e.g. vectors, that contain both scalar and vector finite element variables. The inheritance is a bit similar to that of the [FE libmesh classes](https://libmesh.github.io/doxygen/classlibMesh_1_1FEGenericBase.html) where `FEAbstract` is akin to `MooseVariableFE`, and `FEGenericBase<OutputType>` is akin to `MooseVariableField<OutputType>`

File: framework/src/outputs/Exodus.C [hunk 1]
```diff
@@ -195,6 +195,21 @@ Exodus::outputSetup()
   }
 }
 
+void
+Exodus::outputVectorVariables()
+{
+  // std::vector<std::string> vector(getVectorVariableOutput().begin(),
+  // getVectorVariableOutput().end());
```
Comment (andrsd) [inline framework/src/outputs/Exodus.C]: we do not need this...

File: framework/src/postprocessors/ElementIntegralVariablePostprocessor.C [hunk 2]
```diff
@@ -26,12 +26,18 @@ validParams<ElementIntegralVariablePostprocessor>()
...
 {
-  addMooseVariableDependency(mooseVariable());
+  // Try the scalar version first
+  std::string variable_name = parameters.getMooseType("variable");
+  if (variable_name == "")
+    // When using vector variables, we are only going to use the first one in the list at the
+    // interface level...
+    variable_name = parameters.getVecMooseType("variable")[0];
+
+  addMooseVariableDependency(&_subproblem.getVariable(_tid, variable_name));
```
Comment (andrsd) [inline framework/src/postprocessors/ElementIntegralVariablePostprocessor.C]: These bits that repeat feels like they should have stayed in `MooseVariableInterface` you removed.  Maybe we need `MooseVectorVariableInterface` for the vector-valued systems?

File: modules/navier_stokes/src/userobjects/CNSFVGreenGaussSlopeReconstruction.C [hunk 3]
```diff
@@ -35,11 +35,11 @@ validParams<CNSFVGreenGaussSlopeReconstruction>()
...
-    _rho(getVar("rho", 0)),
-    _rhou(getVar("rhou", 0)),
-    _rhov(isCoupled("rhov") ? getVar("rhov", 0) : nullptr),
-    _rhow(isCoupled("rhow") ? getVar("rhow", 0) : nullptr),
-    _rhoe(getVar("rhoe", 0)),
+    _rho(dynamic_cast<MooseVariable *>(getVar("rho", 0))),
+    _rhou(dynamic_cast<MooseVariable *>(getVar("rhou", 0))),
+    _rhov(isCoupled("rhov") ? dynamic_cast<MooseVariable *>(getVar("rhov", 0)) : nullptr),
+    _rhow(isCoupled("rhow") ? dynamic_cast<MooseVariable *>(getVar("rhow", 0)) : nullptr),
+    _rhoe(dynamic_cast<MooseVariable *>(getVar("rhoe", 0))),
```
Comment (andrsd) [inline modules/navier_stokes/src/userobjects/CNSFVGreenGaussSlopeReconstruction.C]: Yuck!

File: modules/phase_field/include/kernels/CHInterfaceBase.h [hunk 4]
```diff
@@ -67,9 +67,9 @@ CHInterfaceBase<T>::CHInterfaceBase(const InputParameters & parameters)
     _M(getMaterialProperty<T>("mob_name")),
     _dMdc(getMaterialPropertyDerivative<T>("mob_name", _var.name())),
     _d2Mdc2(getMaterialPropertyDerivative<T>("mob_name", _var.name(), _var.name())),
-    _second_u(second()),
-    _second_test(secondTest()),
-    _second_phi(secondPhi()),
+    _second_u(_var.secondSln()),
+    _second_test(_var.secondPhi()),
+    _second_phi(_var.secondPhi()),
```
Comment (andrsd) [inline modules/phase_field/include/kernels/CHInterfaceBase.h]: And there is another reason why `MooseVariableInterface` should have stayed...

File: test/tests/ics/dependency/monomial.i [hunk 5]
```diff
@@ -86,32 +86,36 @@
...
+    penalty = 1000
   [../]
 
   [./right_v]
-    type = DirichletBC
+    type = PenaltyDirichletBC
     variable = v
     boundary = right
     value = 1
+    penalty = 1000
```
Comment (andrsd) [inline test/tests/ics/dependency/monomial.i]: These changes look unrelated to your PR. Why was this changed?

File: test/tests/kernels/vector_fe/tests [hunk 6]
```diff
@@ -0,0 +1,8 @@
+[Tests]
+  [./vector_fe]
+    type = Exodiff
+    input = 'vector_kernel.i'
+    exodiff = 'vector_kernel_out.e'
+    abs_zero = 1e-9
+  [../]
```
Comment (andrsd) [inline test/tests/kernels/vector_fe/tests]: One test rules them all? :-)

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by lindsayad** [comment]
Thanks a lot for the review @andrsd. I hope to get to it soon.

File: framework/src/bcs/IntegratedBC.C [hunk 7]
```diff
@@ -27,65 +27,28 @@ template <>
...
-    CoupleableMooseVariableDependencyIntermediateInterface(this, false),
-    MaterialPropertyInterface(this, boundaryIDs()),
-    _current_elem(_assembly.elem()),
-    _current_elem_volume(_assembly.elemVolume()),
-    _current_side(_assembly.side()),
-    _current_side_elem(_assembly.sideElem()),
-    _current_side_volume(_assembly.sideElemVolume()),
-
+  : IntegratedBCBase(parameters), // False is because this is NOT nodal
+    _var(dynamic_cast<MooseVariable &>(_base_var)),
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/src/bcs/IntegratedBC.C]: @andrsd Would templating the getVariable methods make sense?

[the remark this answers is not in the document]
Comment (andrsd) [reply framework/src/bcs/IntegratedBC.C]: Yes, it would.

[the remark this answers is not in the document]
Comment (andrsd) [reply framework/src/bcs/IntegratedBC.C]: Although you might have an issue in `AuxiliarySystem`, because your PR does not allow vector-valued aux-variables. But, you should be able to use explicit instantiation to prevent people to add/get them.

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/src/bcs/IntegratedBC.C]: What design do you like better: 
1) A single container of MooseVariableBase pointers in the VariableWarehouse with `dynamic_cast` checking in the templated `SystemBase` getter
2) Separate (unordered)maps of MooseVariables and MooseVariableVectors and a templated getter method in the VariableWarehouse that removes the need for dynamic casting in the `SystemBase` getter

[the remark this answers is not in the document]
Comment (andrsd) [reply framework/src/bcs/IntegratedBC.C]: I'd go with 2. We already have normal + scalar variables there, so it is just logical we add vector valued variables there.

File: modules/phase_field/include/kernels/CHInterfaceBase.h (hunk 4 above)
Comment (lindsayad) [reply modules/phase_field/include/kernels/CHInterfaceBase.h]: But it's only 5 more chracters vs. a whole intermediate class? Now the repeated call of `_sys.getVariable(_tid, parameters.get<NonlinearVariableName>("variable"))` (or `_subproblem...`) in all the computing object classes does seem tedious I'll agree

Comment (andrsd) [reply modules/phase_field/include/kernels/CHInterfaceBase.h]: And let's not forget we might hide some smart things in that `MooseVariableInterface` c-tor. Those would get propagated into all interfaces that are using the class...

File: test/tests/ics/dependency/monomial.i (hunk 5 above)
Comment (lindsayad) [reply test/tests/ics/dependency/monomial.i]: With my changes this test errors out. I suppose there are a couple of ways to think about DirichletBC. If we think about it as assigning values for DOFs then I suppose my changes should be compatible with this test as it stands in HEAD. However, if we think about them as assigning values on domain boundaries then I don't think using a DirichletBC makes sense in the context of discontinuous monomials.

**Comment by lindsayad** [comment]
@andrsd Hopefully the commit history looks a bit better now. Thanks for pushing me to explore `fixup` and `autosquash`. Neat tools.

Next thing I'll work on is adding a test demonstrating use of `LAGRANGE_VEC`

File: test/tests/kernels/vector_fe/tests (hunk 6 above)
Comment (lindsayad) [reply test/tests/kernels/vector_fe/tests]: I've added a couple more tests. One demonstrates coupling a vector variable into a standard kernel. The other tests the Jacobian for the former case since we introduce a `_vector_phi` member on top of `_phi`.

I would like to add a LAGRANGE_VEC test, but without too much more work it will have to use integrated boundary conditions. It's currently possible in libmesh to specify dirichlet boundary conditions for LAGRANGE_VEC, but we'll have to add some additional machinery to MOOSE to have it work for us.

**Comment by lindsayad** [comment]
I think this is ready for further review

File: framework/include/utils/MooseTypes.h [hunk 8]
```diff
@@ -71,6 +72,21 @@
 #pragma clang diagnostic pop
 #endif
 
+/**
+ * forward declarations
+ */
+namespace libMesh
+{
+template <typename>
+class VectorValue;
```
Comment (andrsd) [inline framework/include/utils/MooseTypes.h]: Are you not getting this via `#include "libmesh/vector_value.h"`?

File: framework/include/utils/MooseTypes.h [hunk 9]
```diff
@@ -71,6 +72,21 @@
...
+template <typename>
+class TensorValue;
+template <unsigned int, typename>
+class TypeNTensor;
```
Comment (andrsd) [inline framework/include/utils/MooseTypes.h]: `#include "libmesh/tensor_value.h"` and `"type_n_tensor.h"` and then you do not need this.

File: framework/src/bcs/IntegratedBCBase.C [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "IntegratedBCBase.h"
+
+// MOOSE includes
```
Comment (andrsd) [inline framework/src/bcs/IntegratedBCBase.C]: The first one is not MOOSE include? ;-)
Let's not put cpt. Obvious in the code.

File: framework/src/kernels/VectorKernel.C [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+
+#include "VectorKernel.h"
+
+// MOOSE includes
```
Comment (andrsd) [inline framework/src/kernels/VectorKernel.C]: cpt. Obvious.

File: modules/phase_field/include/kernels/CahnHilliardBase.h [hunk 12]
```diff
@@ -100,7 +100,7 @@ CahnHilliardBase<T>::CahnHilliardBase(const InputParameters & parameters)
           &this->template getMaterialPropertyDerivative<Real>("f_name", _var.name(), iname, jname);
     }
 
-    _grad_vars[i + 1] = &(_coupled_moose_vars[i]->gradSln());
+    _grad_vars[i + 1] = &this->_sys.template getFieldVariable<Real>(this->_tid, iname).gradSln();
```
Comment (andrsd) [inline modules/phase_field/include/kernels/CahnHilliardBase.h]: I do not think we want this in user code. Can we provide better API?

File: modules/phase_field/src/kernels/ACInterface.C [hunk 13]
```diff
@@ -41,7 +41,7 @@ ACInterface::ACInterface(const InputParameters & parameters)
   // Get mobility and kappa derivatives and coupled variable gradients
   for (unsigned int i = 0; i < _nvar; ++i)
   {
-    MooseVariable * ivar = _coupled_moose_vars[i];
+    MooseVariable * ivar = dynamic_cast<MooseVariable *>(_coupled_moose_vars[i]);
```
Comment (andrsd) [inline modules/phase_field/src/kernels/ACInterface.C]: in `ACSwitching` you used `MooseVariableFE` and did not have to use `dynamic_cast`. why could not you use it here?

File: test/include/kernels/EFieldAdvection.h [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+/*                   ALL RIGHTS RESERVED                        */
+/*                                                              */
+/*          Prepared by Battelle Energy Alliance, LLC           */
+/*            Under Contract No. DE-AC07-05ID14517              */
+/*            With the U. S. Department of Energy               */
+/*                                                              */
+/*            See COPYRIGHT for full restrictions               */
+/****************************************************************/
+#ifndef EFIELDADVECTION_H_
+#define EFIELDADVECTION_H_
```
Comment (andrsd) [inline test/include/kernels/EFieldAdvection.h]: no trailing `_`

File: test/include/kernels/EFieldAdvection.h [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+{
+public:
+  EFieldAdvection(const InputParameters & parameters);
+
+protected:
+  virtual Real computeQpResidual();
+  virtual Real computeQpJacobian();
+  virtual Real computeQpOffDiagJacobian(unsigned int jvar);
+
+private:
```
Comment (andrsd) [inline test/include/kernels/EFieldAdvection.h]: So whoever inherits from this won't be able to use those?

**Review by andrsd — DISMISSED** [review]

File: framework/src/base/NeighborCoupleable.C [hunk 16]
```diff
@@ -32,9 +32,9 @@ NeighborCoupleable::~NeighborCoupleable() {}
 const VariableValue &
 NeighborCoupleable::coupledNeighborValue(const std::string & var_name, unsigned int comp)
 {
-  MooseVariable * var = getVar(var_name, comp);
+  MooseVariable * var = dynamic_cast<MooseVariable *>(getVar(var_name, comp));
```
Comment (andrsd) [inline framework/src/base/NeighborCoupleable.C]: Are we sure this will be always not NULL?

Comment (andrsd) [reply framework/src/base/NeighborCoupleable.C]: Can we have an API that will give back `MooseVariable *` that will be checked for `!nullptr` and in case if `nullptr`, we get an error message. It would be a lot better than to have not-checked `dynamic_cast`s.

File: framework/include/kernels/KernelBase.h [hunk 17]
```diff
@@ -93,13 +96,26 @@ class KernelBase : public MooseObject,
...
+  virtual MooseVariableFE & variable() = 0;
 
   /// Returns a reference to the SubProblem for which this Kernel is active
-  SubProblem & subProblem();
+  SubProblem & subProblem() { return _subproblem; }
 
   virtual bool isEigenKernel() const { return _eigen_kernel; }
 
+protected:
+  /// Compute this Kernel's contribution to the residual at the current quadrature point
```
[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/kernels/KernelBase.h]: I thought our style was to use triple slash for doxygen comments that fit on a single line and the regular block style for multi-line comments?

File: modules/phase_field/src/kernels/ACInterface.C (hunk 13 above)
Comment (lindsayad) [reply modules/phase_field/src/kernels/ACInterface.C]: We call the `gradSln` method on `ivar` which isn't implemented for the abstract `MooseVariableFE`. We could do a dance where we get the `MooseVariable` using the `_subproblem` or `_sys` getters with `_coupled_moose_vars[i]->name()`. Or I could add another method(s) in `Coupleable` like `getCoupledStandardMooseVars` (and `getCoupledVectorMooseVars`) which would require adding a few more data members as well. I'm feeling lazy but will implement whatever design you think best

File: framework/src/base/NeighborCoupleable.C (hunk 16 above)
Comment (lindsayad) [reply framework/src/base/NeighborCoupleable.C]: `Coupleable::getVar` does this but these eluded my search and destroy of unnecessary casts

File: framework/include/kernels/KernelBase.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (andrsd) [reply framework/include/kernels/KernelBase.h]: IIRC, `///` for variables and `/** */` for methods.

[the remark this answers is not in the document]
Comment (lindsayad) [reply framework/include/kernels/KernelBase.h]: Ah, got ya

File: framework/src/base/NeighborCoupleable.C (hunk 16 above)
Comment (andrsd) [reply framework/src/base/NeighborCoupleable.C]: Right, I missed this was in `NeighborCoupleable`. Easy fix then. 👍

**Comment by lindsayad** [comment]
If we're happy with what's going on here, I will start preparing patches for the Apps...

**Comment by mangerij** [comment]
So when can we do light scattering? :)

**Comment by lindsayad** [comment]
So, I have patches ready for all the failing apps, built, tested, and submitted to each. So unless there's anything else we want changed, I'm considering this ready to go. @andrsd your thoughts?

**Comment by andrsd** [comment]
> @andrsd your thoughts?

Let the world explode!

**Review by andrsd — APPROVED** [review]

**Comment by dschwen** [comment]
@lindsayad, is the divergence not just the trace of the VectorVariableGradient?

**Comment by lindsayad** [comment]
In Cartesian coordinates this is true but not for general coordinate systems IIRC

**Comment by dschwen** [comment]
So, has anyone already tried implementing Maxwell's equations then? 🏓 @tonkmr

**Comment by mangerij** [comment]
I have a few projects in mind but at the moment it is very hard to find the
time.

Would be cool to see how the framework handles light scattering problems
(those typically calculated from FDTD methods). I imagine the materials
system could be useful in defining things like the conductivity and
dielectric tensors.

**Comment by tonkmr** [comment]
I am currently working on a proposal where, if it gets funded, we will start to solve Maxwell's equations

**Comment by lindsayad** [comment]
You often make a sub-selection of maxwells equations you want to solve. The first vector fe test in the framework essentially solves a combination of amperes and faradays laws (combined into a single curl curl equation). Or you can use alternative gauge formulations which use potentials as opposed to the standard form of maxwells equations. There's a new test in #10933 that solves a Coulomb gauge equation, although that formulation really isn't used very often in practice.

**Comment by mangerij** [comment]
Solving Helmholtz equations with anisotropic inhomogeneous dielectric
fields would be ideal. We could even load them in from Ferret.

I don't know if it's even doable. I think its a nightmare of numerics to
say your wave can arbitrarily change speed AND polarization as it passes
from one grid point to the next.

**Comment by cticenhour** [comment]
I am currently working on implementing the Helmholtz equation for electric and magnetic fields in MOOSE as part of my dissertation research. One planned application for the code would be in RF wave heating (ICRH in fusion tokamaks) where we would be using a dielectric tensor to represent the plasma physics (but I haven't gotten there quite yet). @mangerij, I was actually initially thinking of using the materials system for defining this tensor since it seems to have much of what I would need to get started, at first glance.
